File: push/src/main/java/com/xiaomi/channel/commonutils/msa/MdidIdManager.java
package com.xiaomi.channel.commonutils.msa;

import android.content.Context;
import android.os.Looper;
import android.os.SystemClock;
import android.text.TextUtils;
import com.xiaomi.channel.commonutils.android.SystemUtils;
import com.xiaomi.channel.commonutils.logger.MyLog;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/channel/commonutils/msa/MdidIdManager.class */
class MdidIdManager implements IdManager, InvocationHandler {
    private static final String CORE_CLASS_MDID = "com.bun.miitmdid.core.MdidSdk";
    private static final String[][] LISTENER_SUPPLIER_CLASS_ARRAY = {new String[]{"com.bun.supplier.IIdentifierListener", "com.bun.supplier.IdSupplier"}, new String[]{"com.bun.miitmdid.core.IIdentifierListener", "com.bun.miitmdid.supplier.IdSupplier"}};
    private static final String LOG_PREFIX = "mdid:";
    private static final int MAX_RETRY_COUNT = 3;
    private static final int TIME_WAIT_LOCK = 3000;
    private Context mContext;
    private Class mClassMdid = null;
    private Class mClassIIdentifierListener = null;
    private Method mMethodInitSdk = null;
    private Method mMethodGetUDID = null;
    private Method mMethodGetOAID = null;
    private Method mMethodGetVAID = null;
    private Method mMethodGetAAID = null;
    private Method mMethodIsSupported = null;
    private Method mMethodShutDown = null;
    private final Object mLockObj = new Object();
    private volatile int mRetryCount = 0;
    private volatile long mGettingOrGotTime = 0;
    private volatile IdData mIdData = null;

    /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/channel/commonutils/msa/MdidIdManager$IdData.class */
    private class IdData {
        String aaid;
        Boolean isSupport;
        String oaid;
        String udid;
        String vaid;

        private IdData() {
            this.isSupport = null;
            this.udid = null;
            this.oaid = null;
            this.vaid = null;
            this.aaid = null;
        }

        boolean reviseSelf() {
            boolean z = true;
            if (!TextUtils.isEmpty(this.udid) || !TextUtils.isEmpty(this.oaid) || !TextUtils.isEmpty(this.vaid) || !TextUtils.isEmpty(this.aaid)) {
                this.isSupport = true;
            }
            if (this.isSupport == null) {
                z = false;
            }
            return z;
        }
    }

    public MdidIdManager(Context context) {
        this.mContext = context.getApplicationContext();
        initClass(context);
        callInitSdk(context);
    }

    private void callInitSdk(Context context) {
        long jElapsedRealtime = SystemClock.elapsedRealtime();
        long j = -jElapsedRealtime;
        Class cls = this.mClassIIdentifierListener;
        long j2 = j;
        if (cls != null) {
            try {
                ClassLoader classLoader = cls.getClassLoader();
                ClassLoader classLoader2 = classLoader;
                if (classLoader == null) {
                    classLoader2 = context.getClassLoader();
                }
                invokeMethod(this.mMethodInitSdk, this.mClassMdid.newInstance(), context, Proxy.newProxyInstance(classLoader2, new Class[]{this.mClassIIdentifierListener}, this));
                j2 = jElapsedRealtime;
            } catch (Throwable th) {
                outLog("call init sdk error:" + th);
                j2 = j;
            }
        }
        this.mGettingOrGotTime = j2;
    }

    private void cancelWait() {
        synchronized (this.mLockObj) {
            try {
                this.mLockObj.notifyAll();
            } catch (Exception e) {
            }
        }
    }

    private static Method getMethod(Class<?> cls, String str, Class<?>... clsArr) {
        if (cls == null) {
            return null;
        }
        try {
            return cls.getMethod(str, clsArr);
        } catch (Throwable th) {
            return null;
        }
    }

    private void initClass(Context context) {
        Class<?> clsLoadClass = loadClass(context, CORE_CLASS_MDID);
        Class<?> clsLoadClass2 = null;
        Class<?> clsLoadClass3 = null;
        int i = 0;
        while (true) {
            String[][] strArr = LISTENER_SUPPLIER_CLASS_ARRAY;
            if (i >= strArr.length) {
                break;
            }
            String[] strArr2 = strArr[i];
            clsLoadClass2 = loadClass(context, strArr2[0]);
            clsLoadClass3 = loadClass(context, strArr2[1]);
            if (clsLoadClass2 != null && clsLoadClass3 != null) {
                outLog("found class in index " + i);
                break;
            }
            i++;
        }
        this.mClassMdid = clsLoadClass;
        this.mMethodInitSdk = getMethod(clsLoadClass, "InitSdk", Context.class, clsLoadClass2);
        this.mClassIIdentifierListener = clsLoadClass2;
        this.mMethodGetUDID = getMethod(clsLoadClass3, "getUDID", new Class[0]);
        this.mMethodGetOAID = getMethod(clsLoadClass3, "getOAID", new Class[0]);
        this.mMethodGetVAID = getMethod(clsLoadClass3, "getVAID", new Class[0]);
        this.mMethodGetAAID = getMethod(clsLoadClass3, "getAAID", new Class[0]);
        this.mMethodIsSupported = getMethod(clsLoadClass3, "isSupported", new Class[0]);
        this.mMethodShutDown = getMethod(clsLoadClass3, "shutDown", new Class[0]);
    }

    private static <T> T invokeMethod(Method method, Object obj, Object... objArr) {
        if (method == null) {
            return null;
        }
        try {
            T t = (T) method.invoke(obj, objArr);
            if (t != null) {
                return t;
            }
            return null;
        } catch (Throwable th) {
            return null;
        }
    }

    private static boolean isPrimitive(Object obj) {
        return (obj instanceof Boolean) || (obj instanceof Character) || (obj instanceof Byte) || (obj instanceof Short) || (obj instanceof Integer) || (obj instanceof Long) || (obj instanceof Float) || (obj instanceof Double);
    }

    private static Class<?> loadClass(Context context, String str) {
        try {
            return SystemUtils.loadClass(context, str);
        } catch (Throwable th) {
            return null;
        }
    }

    private static void outLog(String str) {
        MyLog.w(LOG_PREFIX + str);
    }

    private void waitAndGettingIfNeed(String str) {
        if (this.mIdData != null) {
            return;
        }
        long j = this.mGettingOrGotTime;
        long jElapsedRealtime = SystemClock.elapsedRealtime() - Math.abs(j);
        int i = this.mRetryCount;
        long j2 = j;
        long jElapsedRealtime2 = jElapsedRealtime;
        if (jElapsedRealtime > 3000) {
            j2 = j;
            jElapsedRealtime2 = jElapsedRealtime;
            if (i < 3) {
                synchronized (this.mLockObj) {
                    j2 = j;
                    jElapsedRealtime2 = jElapsedRealtime;
                    if (this.mGettingOrGotTime == j) {
                        j2 = j;
                        jElapsedRealtime2 = jElapsedRealtime;
                        if (this.mRetryCount == i) {
                            outLog("retry, current count is " + i);
                            this.mRetryCount = this.mRetryCount + 1;
                            callInitSdk(this.mContext);
                            j2 = this.mGettingOrGotTime;
                            jElapsedRealtime2 = SystemClock.elapsedRealtime() - Math.abs(j2);
                        }
                    }
                }
            }
        }
        if (this.mIdData != null || j2 < 0 || jElapsedRealtime2 > 3000 || Looper.myLooper() == Looper.getMainLooper()) {
            return;
        }
        synchronized (this.mLockObj) {
            if (this.mIdData == null) {
                try {
                    outLog(str + " wait...");
                    this.mLockObj.wait(3000L);
                } catch (Exception e) {
                }
            }
        }
    }

    @Override // com.xiaomi.channel.commonutils.msa.IdManager
    public String getAAID() {
        waitAndGettingIfNeed("getAAID");
        return this.mIdData == null ? null : this.mIdData.aaid;
    }

    @Override // com.xiaomi.channel.commonutils.msa.IdManager
    public String getOAID() {
        waitAndGettingIfNeed("getOAID");
        return this.mIdData == null ? null : this.mIdData.oaid;
    }

    @Override // com.xiaomi.channel.commonutils.msa.IdManager
    public String getUDID() {
        waitAndGettingIfNeed("getUDID");
        return this.mIdData == null ? null : this.mIdData.udid;
    }

    @Override // com.xiaomi.channel.commonutils.msa.IdManager
    public String getVAID() {
        waitAndGettingIfNeed("getVAID");
        return this.mIdData == null ? null : this.mIdData.vaid;
    }

    @Override // java.lang.reflect.InvocationHandler
    public Object invoke(Object obj, Method method, Object[] objArr) {
        this.mGettingOrGotTime = SystemClock.elapsedRealtime();
        if (objArr != null) {
            IdData idData = new IdData();
            int length = objArr.length;
            boolean z = false;
            int i = 0;
            while (true) {
                if (i >= length) {
                    break;
                }
                Object obj2 = objArr[i];
                if (obj2 != null && !isPrimitive(obj2)) {
                    idData.udid = (String) invokeMethod(this.mMethodGetUDID, obj2, new Object[0]);
                    idData.oaid = (String) invokeMethod(this.mMethodGetOAID, obj2, new Object[0]);
                    idData.vaid = (String) invokeMethod(this.mMethodGetVAID, obj2, new Object[0]);
                    idData.aaid = (String) invokeMethod(this.mMethodGetAAID, obj2, new Object[0]);
                    idData.isSupport = (Boolean) invokeMethod(this.mMethodIsSupported, obj2, new Object[0]);
                    invokeMethod(this.mMethodShutDown, obj2, new Object[0]);
                    if (idData.reviseSelf()) {
                        StringBuilder sb = new StringBuilder();
                        sb.append("has get succ, check duplicate:");
                        if (this.mIdData != null) {
                            z = true;
                        }
                        sb.append(z);
                        outLog(sb.toString());
                        synchronized (MdidIdManager.class) {
                            try {
                                if (this.mIdData == null) {
                                    this.mIdData = idData;
                                }
                            } catch (Throwable th) {
                                throw th;
                            }
                        }
                    }
                }
                i++;
            }
        }
        cancelWait();
        return null;
    }

    @Override // com.xiaomi.channel.commonutils.msa.IdManager
    public boolean isAllowOAID() {
        return true;
    }

    @Override // com.xiaomi.channel.commonutils.msa.IdManager
    public boolean isSupported() {
        waitAndGettingIfNeed("isSupported");
        return this.mIdData != null && Boolean.TRUE.equals(this.mIdData.isSupport);
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/channel/commonutils/msa/MdidIdManager.kt
package com.xiaomi.channel.commonutils.msa

import android.content.Context
import android.os.Looper
import android.os.SystemClock
import android.text.TextUtils
import com.xiaomi.channel.commonutils.android.SystemUtils
import com.xiaomi.channel.commonutils.logger.MyLog
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy
import kotlin.math.abs

/*
 * Current override reference: com.xiaomi.xmsf 0.3.17-20260410000745 (versionCode 1003003000),
 * base.apk sha256 f3d72b6f5e1427ceecd3147a051d58e4dc95bb528397d486658e01cad9f7e590,
 * JADX path: com.xiaomi.xmsf/current/base/sources/com/xiaomi/channel/commonutils/msa/MdidIdManager.java
 */
class MdidIdManager(context: Context) : IdManager, InvocationHandler {
    private val mContext: Context = context.applicationContext
    private var mClassMdid: Class<*>? = null
    private var mClassIIdentifierListener: Class<*>? = null
    private var mMethodInitSdk: Method? = null
    private var mMethodGetUDID: Method? = null
    private var mMethodGetOAID: Method? = null
    private var mMethodGetVAID: Method? = null
    private var mMethodGetAAID: Method? = null
    private var mMethodIsSupported: Method? = null
    private var mMethodShutDown: Method? = null
    private val mLockObj = Object()
    @Volatile
    private var mRetryCount = 0
    @Volatile
    private var mGettingOrGotTime = 0L
    @Volatile
    private var mIdData: IdData? = null

    private inner class IdData {
        var aaid: String? = null
        var isSupport: Boolean? = null
        var oaid: String? = null
        var udid: String? = null
        var vaid: String? = null

        fun reviseSelf(): Boolean {
            if (
                !TextUtils.isEmpty(udid) ||
                !TextUtils.isEmpty(oaid) ||
                !TextUtils.isEmpty(vaid) ||
                !TextUtils.isEmpty(aaid)
            ) {
                isSupport = true
            }
            return isSupport != null
        }
    }

    init {
        initClass(context)
        callInitSdk(context)
    }

    private fun callInitSdk(context: Context) {
        val elapsedRealtime = SystemClock.elapsedRealtime()
        var result = -elapsedRealtime
        val listenerClass = mClassIIdentifierListener
        if (listenerClass != null) {
            try {
                var classLoader = listenerClass.classLoader
                if (classLoader == null) {
                    classLoader = context.classLoader
                }
                invokeMethod(
                    mMethodInitSdk,
                    mClassMdid?.getDeclaredConstructor()?.newInstance(),
                    context,
                    Proxy.newProxyInstance(classLoader, arrayOf(listenerClass), this)
                )
                result = elapsedRealtime
            } catch (throwable: Throwable) {
                outLog("call init sdk error:$throwable")
            }
        }
        mGettingOrGotTime = result
    }

    private fun cancelWait() {
        synchronized(mLockObj) {
            try {
                mLockObj.notifyAll()
            } catch (e: Exception) {
                // ignore
            }
        }
    }

    private fun initClass(context: Context) {
        val mdidClass = loadClass(context, CORE_CLASS_MDID)
        var listenerClass: Class<*>? = null
        var supplierClass: Class<*>? = null
        for (i in LISTENER_SUPPLIER_CLASS_ARRAY.indices) {
            val classes = LISTENER_SUPPLIER_CLASS_ARRAY[i]
            listenerClass = loadClass(context, classes[0])
            supplierClass = loadClass(context, classes[1])
            if (listenerClass != null && supplierClass != null) {
                outLog("found class in index $i")
                break
            }
        }
        mClassMdid = mdidClass
        mMethodInitSdk = getMethod(mdidClass, "InitSdk", Context::class.java, listenerClass)
        mClassIIdentifierListener = listenerClass
        mMethodGetUDID = getMethod(supplierClass, "getUDID")
        mMethodGetOAID = getMethod(supplierClass, "getOAID")
        mMethodGetVAID = getMethod(supplierClass, "getVAID")
        mMethodGetAAID = getMethod(supplierClass, "getAAID")
        mMethodIsSupported = getMethod(supplierClass, "isSupported")
        mMethodShutDown = getMethod(supplierClass, "shutDown")
    }

    private fun waitAndGettingIfNeed(str: String) {
        if (mIdData != null) {
            return
        }
        val start = mGettingOrGotTime
        var effectiveStart = start
        var elapsed = SystemClock.elapsedRealtime() - abs(start)
        val retryCount = mRetryCount
        if (elapsed > TIME_WAIT_LOCK && retryCount < MAX_RETRY_COUNT) {
            synchronized(mLockObj) {
                if (mGettingOrGotTime == start && mRetryCount == retryCount) {
                    outLog("retry, current count is $retryCount")
                    mRetryCount += 1
                    callInitSdk(mContext)
                    effectiveStart = mGettingOrGotTime
                    elapsed = SystemClock.elapsedRealtime() - abs(effectiveStart)
                }
            }
        }
        if (
            mIdData != null ||
            effectiveStart < 0 ||
            elapsed > TIME_WAIT_LOCK ||
            Looper.myLooper() == Looper.getMainLooper()
        ) {
            return
        }
        synchronized(mLockObj) {
            if (mIdData == null) {
                try {
                    outLog("$str wait...")
                    mLockObj.wait(TIME_WAIT_LOCK.toLong())
                } catch (e: Exception) {
                    // ignore
                }
            }
        }
    }

    override fun getAAID(): String? {
        waitAndGettingIfNeed("getAAID")
        return mIdData?.aaid
    }

    override fun getOAID(): String? {
        waitAndGettingIfNeed("getOAID")
        return mIdData?.oaid
    }

    override fun getUDID(): String? {
        waitAndGettingIfNeed("getUDID")
        return mIdData?.udid
    }

    override fun getVAID(): String? {
        waitAndGettingIfNeed("getVAID")
        return mIdData?.vaid
    }

    override fun invoke(proxy: Any?, method: Method?, args: Array<Any?>?): Any? {
        mGettingOrGotTime = SystemClock.elapsedRealtime()
        if (args != null) {
            val idData = IdData()
            for (arg in args) {
                if (arg != null && !isPrimitive(arg)) {
                    idData.udid = invokeMethod(mMethodGetUDID, arg) as? String
                    idData.oaid = invokeMethod(mMethodGetOAID, arg) as? String
                    idData.vaid = invokeMethod(mMethodGetVAID, arg) as? String
                    idData.aaid = invokeMethod(mMethodGetAAID, arg) as? String
                    idData.isSupport = invokeMethod(mMethodIsSupported, arg) as? Boolean
                    invokeMethod(mMethodShutDown, arg)
                    if (idData.reviseSelf()) {
                        outLog("has get succ, check duplicate:${mIdData != null}")
                        synchronized(MdidIdManager::class.java) {
                            if (mIdData == null) {
                                mIdData = idData
                            }
                        }
                    }
                }
            }
        }
        cancelWait()
        return null
    }

    override fun isAllowOAID(): Boolean = true

    override fun isSupported(): Boolean {
        waitAndGettingIfNeed("isSupported")
        return mIdData != null && java.lang.Boolean.TRUE == mIdData?.isSupport
    }

    companion object {
        private const val CORE_CLASS_MDID = "com.bun.miitmdid.core.MdidSdk"
        private val LISTENER_SUPPLIER_CLASS_ARRAY = arrayOf(
            arrayOf("com.bun.supplier.IIdentifierListener", "com.bun.supplier.IdSupplier"),
            arrayOf("com.bun.miitmdid.core.IIdentifierListener", "com.bun.miitmdid.supplier.IdSupplier")
        )
        private const val LOG_PREFIX = "mdid:"
        private const val MAX_RETRY_COUNT = 3
        private const val TIME_WAIT_LOCK = 3000L

        private fun getMethod(cls: Class<*>?, name: String, vararg parameterTypes: Class<*>?): Method? {
            if (cls == null || parameterTypes.any { it == null }) {
                return null
            }
            return try {
                @Suppress("UNCHECKED_CAST")
                cls.getMethod(name, *(parameterTypes as Array<Class<*>>))
            } catch (throwable: Throwable) {
                null
            }
        }

        private fun invokeMethod(method: Method?, obj: Any?, vararg args: Any?): Any? {
            if (method == null) {
                return null
            }
            return try {
                method.invoke(obj, *args)
            } catch (throwable: Throwable) {
                null
            }
        }

        private fun isPrimitive(obj: Any): Boolean {
            return obj is Boolean ||
                obj is Char ||
                obj is Byte ||
                obj is Short ||
                obj is Int ||
                obj is Long ||
                obj is Float ||
                obj is Double
        }

        private fun loadClass(context: Context, str: String): Class<*>? {
            return try {
                SystemUtils.loadClass(context, str)
            } catch (throwable: Throwable) {
                null
            }
        }

        private fun outLog(str: String) {
            MyLog.w(LOG_PREFIX + str)
        }
    }
}
